### ISSUE TITLE
Input/output errors and QAT dc_fails

### ISSUE THREAD

**Issue by AGI-chandler** [openzfs/zfs#9784]
### System information

Type | Version/Name
 --- | --- 
Distribution Name	| CentOS
Distribution Version	| 6.10
Linux Kernel	| 2.6.32-754.23.1.el6
Architecture	| x86_64
ZFS Version	| 0.8.2-1
SPL Version	| 0.8.2-1

### Describe the problem you're observing
Getting some Input/output errors when trying to read some data.  "zpool status" shows all disks are ONLINE and "No known data errors", last scrub just finished earlier today.  Noticed that the dc_fails counter is increasing when the I/O errors happen, so I'm guessing it's related to the QAT code in ZFS or the QAT card itself?  The QAT card status is:
```
# /etc/init.d/qat_service status
Checking status of all devices.
There is 3 QAT acceleration device(s) in the system:
 qat_dev0 - type: c6xx,  inst_id: 0,  node_id: 1,  bsf: 0000:84:00.0,  #accel: 5 #engines: 10 state: up
 qat_dev1 - type: c6xx,  inst_id: 1,  node_id: 1,  bsf: 0000:85:00.0,  #accel: 5 #engines: 10 state: up
 qat_dev2 - type: c6xx,  inst_id: 2,  node_id: 1,  bsf: 0000:86:00.0,  #accel: 5 #engines: 10 state: up
#
``` 
and there aren't any errors popping up in the system log when the I/O errors happen.  I could open a case with Intel if you think it would be helpful?

### Describe how to reproduce the problem
```
$ cat /proc/spl/kstat/zfs/qat
19 1 0x01 17 4624 361599045964 3369212093252152
name                            type data
comp_requests                   4    284279855
comp_total_in_bytes             4    35765197023232
comp_total_out_bytes            4    12613266941171
decomp_requests                 4    250580310
decomp_total_in_bytes           4    12703419899870
decomp_total_out_bytes          4    31075581360640
**dc_fails                        4    5804**
encrypt_requests                4    0
encrypt_total_in_bytes          4    0
encrypt_total_out_bytes         4    0
decrypt_requests                4    0
decrypt_total_in_bytes          4    0
decrypt_total_out_bytes         4    0
crypt_fails                     4    0
cksum_requests                  4    1396
cksum_total_in_bytes            4    155779072
cksum_fails                     4    0 

$ rsync -avP X /tmp
sending incremental file list
X
  2444094260 100%   49.18MB/s    0:00:47 (xfer#1, to-check=0/1)
rsync: read errors mapping "X": Input/output error (5)
WARNING: X failed verification -- update discarded (will try again).
X
  2444094260 100%   50.98MB/s    0:00:45 (xfer#2, to-check=0/1)
rsync: read errors mapping "X": Input/output error (5)
ERROR: X failed verification -- update discarded.

sent 4888785353 bytes  received 52 bytes  51733178.89 bytes/sec
total size is 2444094260  speedup is 0.50
rsync error: some files/attrs were not transferred (see previous errors) (code 23) at main.c(1039) [sender=3.0.6]

$ cat /proc/spl/kstat/zfs/qat
19 1 0x01 17 4624 361599045964 3369334123310545
name                            type data
comp_requests                   4    284279939
comp_total_in_bytes             4    35765206926336
comp_total_out_bytes            4    12613268958496
decomp_requests                 4    250617594
decomp_total_in_bytes           4    12707033557470
decomp_total_out_bytes          4    31080467294208
**dc_fails                        4    5808**
encrypt_requests                4    0
encrypt_total_in_bytes          4    0
encrypt_total_out_bytes         4    0
decrypt_requests                4    0
decrypt_total_in_bytes          4    0
decrypt_total_out_bytes         4    0
crypt_fails                     4    0
cksum_requests                  4    1396
cksum_total_in_bytes            4    155779072
cksum_fails                     4    0
$ 
```

### Include any warning/errors/backtraces from the system logs

There are no warning/errors/backtraces from the system logs, just the dc_fails is incremented

Thanks

**Comment by luki-mbi** [comment]
Normally when the QAT request fails, ZFS falls back to the CPU. This is usually very rare on our production systems, maybe like 1 : 100M requests, and an error is logged to dmesg by the QAT driver:

`[5344115.233436] dcCompression_ProcessCallback() - : The slice hang is detected on the compression slice`

My guess is that in your case the data for some reason is corrupt and cannot be decompressed by the QAT or the CPU gzip implementation; the checksum (post-compression) is can still be valid. Just a guess. I hope someone else can chime in on that.

**Comment by AGI-admin** [comment]
If the data is corrupt then that is also concerning, since there's no previous indication of corruption.  This is a new ZFS volume and the data was copied from a backup with rsync.  I can check the secondary backups to see if restoring these files has any effect.

**Comment by AGI-admin** [comment]
Restoring the file from the secondary backup has worked fine, so I guess the data was corrupted somehow.

**Comment by luki-mbi** [comment]
That is worrisome. I hope someone can chime in an elaborate what can be causing such corruption.
Possibly the data got corrupted during or after compression (by the QAT?), and the checksum was computed for the corrupted data. So to ZFS all looks fine, the data are readable and the checksum matches. It just cannot be decompressed and results in an I/O error.

Are you sure the hardware is all OK? No bus or memory errors?

**Comment by AGI-admin** [comment]
I've just looked over last week's system log files, and there aren't any messages about memory errors or other errors.  All the scrubs, consistency checks, patrol reads, and verifications have completed successfully.  From my perspective, the machine is running perfectly.

**Comment by AGI-admin** [comment]
Still getting read errors from bacula when it's trying to read (backup) the same files I just restored (from its backup from several months ago), the "secondary backup" I was speaking about previously.  For example,
`29-Dec 16:00 bacula-fd JobId 28943: Error: Read error on file O_sativa117425/canu1st/trimming/canu1st.ovlStore/0017. ERR=Input/output error
`

rsync has an i/o error as well,
```
$ rsync -avP O_sativa117425/canu1st/trimming/canu1st.ovlStore/0017 /tmp
sending incremental file list
0017
  2485036820 100%   85.53MB/s    0:00:27 (xfer#1, to-check=0/1)
rsync: read errors mapping "O_sativa117425/canu1st/trimming/canu1st.ovlStore/0017": Input/output error (5)

sent 2485340260 bytes  received 31 bytes  84248823.42 bytes/sec
total size is 2485036820  speedup is 1.00
rsync error: some files/attrs were not transferred (see previous errors) (code 23) at main.c(1039) [sender=3.0.6]
$ 
```

  But it's not all the files.  For example the "X" file from yesterday that I restored from backup is still reading fine.  

dc_fails is up to 8905 now.... nothing new in dmesg or the syslog

**Comment by AGI-admin** [comment]
I'm trying to restore the same files from the backup again, and noticing that dc_fails is steadily going up, it's over 9000 now and increasing.  So I would say at this point it seems to be a problem with ZFS sending the data to the QAT card or something, as well as reading data with the QAT card.

Here is one of the files causing problems, restored from the backup -> our web server (they are all 2.4 GB, sorry): [O_nivara/canu2nd/trimming/O_nivara.ovlStore/0022](https://www.genome.arizona.edu/0022)  @luki-mbi perhaps you could see if your QAT setup has problems with this file?

I could try to stop the qat_service too and see what happens, but that will slow down the system too much so it isn't a viable work-around.

**Comment by luki-mbi** [comment]
Sees to work fine for me.
- Original file in /net/scratch (no compression)
- Copy in /projects, compressed with QAT (using a C627 chipset)
- Copy in /archive, compressed with QAT (using a DH8950 PCIe card)

```
# md5sum /net/scratch/lugo/0022 /projects/0022 /archive/lugo/0022 
d27eb19aef7965f71ad75dedcf4e155b  /net/scratch/lugo/0022
d27eb19aef7965f71ad75dedcf4e155b  /projects/0022
d27eb19aef7965f71ad75dedcf4e155b  /archive/lugo/0022

# du -sh /net/scratch/lugo/0022 /projects/0022 /archive/lugo/0022 
2.4G	/net/scratch/lugo/0022
1.9G	/projects/0022
1.8G	/archive/lugo/0022

# dd if=0022  of=/dev/null bs=1M
2366+1 records in
2366+1 records out
2481227080 bytes (2.5 GB, 2.3 GiB) copied, 6.09046 s, 407 MB/s
```

No data errors, and performance is decent too for a single thread (~400 MB/sec).

**Comment by AGI-admin** [comment]
So our QAT card is broken then I guess?  Are you also using ZFS 0.8.2?  You compiled it from source using --with-qat right?

**Comment by behlendorf** [comment]
@AGI-admin I've identified the issue and have opened #9788 with the fix.  The change is straight forward, but unfortunately I don't have easy access to a QAT accelerator so I wasn't able to test it myself.

The issue is that when memory on the system becomes highly fragmented some QAT related memory allocations may fail.  Those failures were not being reported to the higher layers of ZFS which prevented the software fallback from handling the compression.  This fits with the logs you posted which clearly show that memory was highly fragmented on your system when writing these blocks.  i.e.
```
nfsd: page allocation failure. order:4, mode:0x20
```

**Comment by AGI-admin** [comment]
Thanks @behlendorf ... but is it normal to see dc_fails increasing without any "page allocation failure" messages?  Because those page allocation failures were happening a couple weeks ago, but not lately when these I/O errors are still happening....

**Comment by behlendorf** [comment]
The allocation failure messages _might_ have been suppressed by the kernel, but yes I'd have expected to see at least one.  You may also want to check the QAT accelerator firmware counters, they may indicate there's an issue on the hardware side.  If the QAT were to return success when in fact there was a failure, you'd see a similar issue.

```
/sys/kernel/debug/qat_*/fw_counters
```

**Comment by AGI-admin** [comment]
We don't have any of those counters, /sys/kernel/debug is empty,

```
$ cat /sys/kernel/debug/qat_*/fw_counters
cat: /sys/kernel/debug/qat_*/fw_counters: No such file or directory
$ ls /sys/kernel/debug/
$ 
```

**Comment by behlendorf** [comment]
You'll need to mount `debugfs` manually, here's a [link](https://access.redhat.com/documentation/en-US/Red_Hat_Enterprise_MRG/2/html/Realtime_Tuning_Guide/Mount_debugfs.html) to the relevant documentation.  To summarize:

```mount -t debugfs nodev /sys/kernel/debug```

**Comment by AGI-admin** [comment]
OK thanks here's what we have there now, anything in particular to look for?

```
# cat /sys/kernel/debug/qat_*/fw_counters
+------------------------------------------------+
| FW Statistics for Qat Device                   |
+------------------------------------------------+
| Firmware Requests [AE  0]:         12149803939 |
| Firmware Responses[AE  0]:         12149803939 |
+------------------------------------------------+
| Firmware Requests [AE  1]:         11501471210 |
| Firmware Responses[AE  1]:         11501471210 |
+------------------------------------------------+
| Firmware Requests [AE  2]:         12149817476 |
| Firmware Responses[AE  2]:         12149817476 |
+------------------------------------------------+
| Firmware Requests [AE  3]:         11501576382 |
| Firmware Responses[AE  3]:         11501576382 |
+------------------------------------------------+
| Firmware Requests [AE  4]:         12148747996 |
| Firmware Responses[AE  4]:         12148747996 |
+------------------------------------------------+
| Firmware Requests [AE  5]:         11501307552 |
| Firmware Responses[AE  5]:         11501307552 |
+------------------------------------------------+
| Firmware Requests [AE  6]:         12145310500 |
| Firmware Responses[AE  6]:         12145310500 |
+------------------------------------------------+
| Firmware Requests [AE  7]:         11500510194 |
| Firmware Responses[AE  7]:         11500510194 |
+------------------------------------------------+
| Firmware Requests [AE  8]:         12142185315 |
| Firmware Responses[AE  8]:         12142185315 |
+------------------------------------------------+
| Firmware Requests [AE  9]:         11499848788 |
| Firmware Responses[AE  9]:         11499848788 |
+------------------------------------------------+
+------------------------------------------------+
| FW Statistics for Qat Device                   |
+------------------------------------------------+
| Firmware Requests [AE  0]:           625652871 |
| Firmware Responses[AE  0]:           625652871 |
+------------------------------------------------+
| Firmware Requests [AE  1]:           586996627 |
| Firmware Responses[AE  1]:           586996627 |
+------------------------------------------------+
| Firmware Requests [AE  2]:           625656789 |
| Firmware Responses[AE  2]:           625656789 |
+------------------------------------------------+
| Firmware Requests [AE  3]:           587004550 |
| Firmware Responses[AE  3]:           587004550 |
+------------------------------------------------+
| Firmware Requests [AE  4]:           625588159 |
| Firmware Responses[AE  4]:           625588159 |
+------------------------------------------------+
| Firmware Requests [AE  5]:           586988266 |
| Firmware Responses[AE  5]:           586988266 |
+------------------------------------------------+
| Firmware Requests [AE  6]:           625391062 |
| Firmware Responses[AE  6]:           625391062 |
+------------------------------------------------+
| Firmware Requests [AE  7]:           586948068 |
| Firmware Responses[AE  7]:           586948068 |
+------------------------------------------------+
| Firmware Requests [AE  8]:           625196343 |
| Firmware Responses[AE  8]:           625196343 |
+------------------------------------------------+
| Firmware Requests [AE  9]:           586918530 |
| Firmware Responses[AE  9]:           586918530 |
+------------------------------------------------+
+------------------------------------------------+
| FW Statistics for Qat Device                   |
+------------------------------------------------+
| Firmware Requests [AE  0]:            18575554 |
| Firmware Responses[AE  0]:            18575554 |
+------------------------------------------------+
| Firmware Requests [AE  1]:            18575271 |
| Firmware Responses[AE  1]:            18575271 |
+------------------------------------------------+
| Firmware Requests [AE  2]:            18574983 |
| Firmware Responses[AE  2]:            18574983 |
+------------------------------------------------+
| Firmware Requests [AE  3]:            18575350 |
| Firmware Responses[AE  3]:            18575350 |
+------------------------------------------------+
| Firmware Requests [AE  4]:            18575441 |
| Firmware Responses[AE  4]:            18575441 |
+------------------------------------------------+
| Firmware Requests [AE  5]:            18574861 |
| Firmware Responses[AE  5]:            18574861 |
+------------------------------------------------+
| Firmware Requests [AE  6]:            18575357 |
| Firmware Responses[AE  6]:            18575357 |
+------------------------------------------------+
| Firmware Requests [AE  7]:            18575451 |
| Firmware Responses[AE  7]:            18575451 |
+------------------------------------------------+
| Firmware Requests [AE  8]:            18574719 |
| Firmware Responses[AE  8]:            18574719 |
+------------------------------------------------+
| Firmware Requests [AE  9]:            18575148 |
| Firmware Responses[AE  9]:            18575148 |
+------------------------------------------------+
# 
```

**Comment by behlendorf** [comment]
Good question.  I was hoping it would be more self evident, the [documentation](https://01.org/sites/default/files/downloads/intelr-quickassist-technology/336210qatswprogguiderev004.pdf) doesn't appear to describe the individual counters in section 3.7.  If you're willing to patch the ZFS code and rebuild we can patch the code to log the exact error.

**Comment by AGI-admin** [comment]
Yes I'm trying to rebuild it right now with #9788

**Comment by AGI-admin** [comment]
@behlendorf  I'm getting an error when compiling qat_compress.c now,

```
/usr/local/src/zfs-0.8.2/module/zfs/qat_compress.c:544: error: expected declaration specifiers or ‘...’ before ‘zfs_kernel_param_t’
/usr/local/src/zfs-0.8.2/module/zfs/qat_compress.c: In function ‘param_set_qat_compress’:
/usr/local/src/zfs-0.8.2/module/zfs/qat_compress.c:547: error: ‘kp’ undeclared (first use in this function)
/usr/local/src/zfs-0.8.2/module/zfs/qat_compress.c:547: error: (Each undeclared identifier is reported only once
/usr/local/src/zfs-0.8.2/module/zfs/qat_compress.c:547: error: for each function it appears in.)
/usr/local/src/zfs-0.8.2/module/zfs/qat_compress.c: At top level:
/usr/local/src/zfs-0.8.2/module/zfs/qat_compress.c:565: warning: initialization from incompatible pointer type
```

**Comment by behlendorf** [comment]
@AGI-admin it looks like you may need to re-run `./autogen.sh` and `./configure` against the intended kernel.  For some reason, the wrong kernel interface was used.  There _might_ be a typo in the PR but I didn't see anything when visually inspecting it.

**Comment by AGI-admin** [comment]
yeah i did that, after decompressing the source archive... I have to change `#include <sys/qat.h>` to `#include "qat.h"`, but otherwise it builds fine with the old qat_compress.c.

**Comment by AGI-admin** [comment]
ugh this is such a mess, i think it's back to this commit, https://github.com/zfsonlinux/zfs/issues/9268

**Comment by behlendorf** [comment]
Ahh yes, sorry I'd forgotten about that issue.  We'll make sure it gets included in the next tag.

**Comment by AGI-admin** [comment]
Actually i think it was already merged, since the code matches, so not sure why the above error is happening :(

**Comment by AGI-admin** [comment]
So is this bug corrupting our data?  If so, is it just the data that we happen to see I/O errors with?  In that case we can restore the individual files from our backup....

**Comment by AGI-admin** [comment]
Hello @behlendorf , I applied the patch and recompiled ZFS and still getting input/output errors :(  I restored files from backup and still can't read it.  For example,

```
tmp# cat /proc/spl/kstat/zfs/qat
19 1 0x01 17 4624 1271105125637 22259398341141
name                            type data
comp_requests                   4    34962
comp_total_in_bytes             4    4335009280
comp_total_out_bytes            4    2155182815
decomp_requests                 4    249190
decomp_total_in_bytes           4    15959668224
decomp_total_out_bytes          4    32395348992
dc_fails                        4    19
encrypt_requests                4    0
encrypt_total_in_bytes          4    0
encrypt_total_out_bytes         4    0
decrypt_requests                4    0
decrypt_total_in_bytes          4    0
decrypt_total_out_bytes         4    0
crypt_fails                     4    0
cksum_requests                  4    550
cksum_total_in_bytes            4    63078400
cksum_fails                     4    0

# rsync -avP O_sativa117425/canu1st/trimming/canu1st.ovlStore/0017 /tmp
sending incremental file list
0017
  2485036820 100%  106.94MB/s    0:00:22 (xfer#1, to-check=0/1)
rsync: read errors mapping "O_sativa117425/canu1st/trimming/canu1st.ovlStore/0017": Input/output error (5)
WARNING: 0017 failed verification -- update discarded (will try again).
0017
  2485036820 100%  102.12MB/s    0:00:23 (xfer#2, to-check=0/1)
rsync: read errors mapping "O_sativa117425/canu1st/trimming/canu1st.ovlStore/0017": Input/output error (5)
ERROR: 0017 failed verification -- update discarded.

sent 4970680473 bytes  received 52 bytes  109245725.82 bytes/sec
total size is 2485036820  speedup is 0.50
rsync error: some files/attrs were not transferred (see previous errors) (code 23) at main.c(1039) [sender=3.0.6]
# cat /proc/spl/kstat/zfs/qat
19 1 0x01 17 4624 1271105125637 22318995971136
name                            type data
comp_requests                   4    34984
comp_total_in_bytes             4    4337058304
comp_total_out_bytes            4    2155334849
decomp_requests                 4    286905
decomp_total_in_bytes           4    19604866560
decomp_total_out_bytes          4    37337207296
dc_fails                        4    27
encrypt_requests                4    0
encrypt_total_in_bytes          4    0
encrypt_total_out_bytes         4    0
decrypt_requests                4    0
decrypt_total_in_bytes          4    0
decrypt_total_out_bytes         4    0
crypt_fails                     4    0
cksum_requests                  4    550
cksum_total_in_bytes            4    63078400
cksum_fails                     4    0
# 
```

Then I delete this 0017 file and restore from backup from 2019-08-26 (long before ZFS and QAT were installed in the system), dc_fails increased again, and the file is still unreadable,

```
# cat /proc/spl/kstat/zfs/qat
19 1 0x01 17 4624 1271105125637 22716806520828
name                            type data
comp_requests                   4    54054
comp_total_in_bytes             4    6833537024
comp_total_out_bytes            4    3983695013
decomp_requests                 4    287172
decomp_total_in_bytes           4    19614187520
decomp_total_out_bytes          4    37369810432
dc_fails                        4    34
encrypt_requests                4    0
encrypt_total_in_bytes          4    0
encrypt_total_out_bytes         4    0
decrypt_requests                4    0
decrypt_total_in_bytes          4    0
decrypt_total_out_bytes         4    0
crypt_fails                     4    0
cksum_requests                  4    550
cksum_total_in_bytes            4    63078400
cksum_fails                     4    0

# rsync -avP O_sativa117425/canu1st/trimming/canu1st.ovlStore/0017 /tmp
sending incremental file list
0017
  2485036820 100%  112.86MB/s    0:00:20 (xfer#1, to-check=0/1)
rsync: read errors mapping "O_sativa117425/canu1st/trimming/canu1st.ovlStore/0017": Input/output error (5)
WARNING: 0017 failed verification -- update discarded (will try again).
0017
  2485036820 100%  105.85MB/s    0:00:22 (xfer#2, to-check=0/1)
rsync: read errors mapping "O_sativa117425/canu1st/trimming/canu1st.ovlStore/0017": Input/output error (5)
ERROR: 0017 failed verification -- update discarded.

sent 4970680473 bytes  received 52 bytes  111700685.96 bytes/sec
total size is 2485036820  speedup is 0.50
rsync error: some files/attrs were not transferred (see previous errors) (code 23) at main.c(1039) [sender=3.0.6]

# cat /proc/spl/kstat/zfs/qat
19 1 0x01 17 4624 1271105125637 22814570223417
name                            type data
comp_requests                   4    54095
comp_total_in_bytes             4    6837144576
comp_total_out_bytes            4    3983944970
decomp_requests                 4    325519
decomp_total_in_bytes           4    23296637440
decomp_total_out_bytes          4    42393592832
dc_fails                        4    44
encrypt_requests                4    0
encrypt_total_in_bytes          4    0
encrypt_total_out_bytes         4    0
decrypt_requests                4    0
decrypt_total_in_bytes          4    0
decrypt_total_out_bytes         4    0
crypt_fails                     4    0
cksum_requests                  4    550
cksum_total_in_bytes            4    63078400
cksum_fails                     4    0
# 

```

**Comment by behlendorf** [comment]
@AGI-admin thanks for testing the proposed fix.  It sounds like while we fixed one issue it's not the one you're encountering.  I'll reopen this issue while we continue to investigate.

**Comment by aqw** [comment]
@behlendorf We are encountering something that looks to be the same bug (though please let me know if you'd like me to file a separate bug report).

I am running ZFS `0.8.3-1~bpo10+1` (so it should include #9788) on Debian buster with QAT Engine `qat1.7.l.4.9.0-00008`. 

There is one file which throws an `Input/output error` when we attempt to read it. However, a `zpool scrub` succeeds and `zpool status` reports a healthy pool.

However, `dc_fail`  shows that there is an error during decompression. This goes up by one each time I try to read the file.
```
❱ cat /proc/spl/kstat/zfs/qat
19 1 0x01 17 4624 8085381884 1267563807819134
name                            type data
comp_requests                   4    317346892
comp_total_in_bytes             4    41476450213888
comp_total_out_bytes            4    573639004432
decomp_requests                 4    10583940
decomp_total_in_bytes           4    528333138499
decomp_total_out_bytes          4    1305853530624
dc_fails                        4    40
encrypt_requests                4    0
encrypt_total_in_bytes          4    0
encrypt_total_out_bytes         4    0
decrypt_requests                4    0
decrypt_total_in_bytes          4    0
decrypt_total_out_bytes         4    0
crypt_fails                     4    0
cksum_requests                  4    13183364
cksum_total_in_bytes            4    54295887872
cksum_fails                     4    0
```

The QAT card itself also sees the issue:
```
❱ cat /sys/kernel/debug/qat_c6xx_0000:48:00.0/cnv_errors                                                                                 1 !
+-----------------------------------------------------------------+
|             CNV Error Freq Statistics for Qat Device            |
+-----------------------------------------------------------------+
|[AE  0]: TotalErrors:     1 : LastError: Decomp Error   [   -7]  |
+-----------------------------------------------------------------+
|[AE  1]: TotalErrors:     0 : LastError: No Error       [    0]  |
+-----------------------------------------------------------------+
|[AE  2]: TotalErrors:     2 : LastError: Decomp Error   [   -7]  |
+-----------------------------------------------------------------+
|[AE  3]: TotalErrors:     0 : LastError: No Error       [    0]  |
+-----------------------------------------------------------------+
|[AE  4]: TotalErrors:     1 : LastError: Decomp Error   [   -7]  |
+-----------------------------------------------------------------+
|[AE  5]: TotalErrors:     1 : LastError: Decomp Error   [   -7]  |
+-----------------------------------------------------------------+
|[AE  6]: TotalErrors:     0 : LastError: No Error       [    0]  |
+-----------------------------------------------------------------+
|[AE  7]: TotalErrors:     2 : LastError: Decomp Error   [   -7]  |
+-----------------------------------------------------------------+
|[AE  8]: TotalErrors:     1 : LastError: Decomp Error   [   -7]  |
+-----------------------------------------------------------------+
|[AE  9]: TotalErrors:     0 : LastError: No Error       [    0]  |
+-----------------------------------------------------------------+
```

And a line in dmesg is quite persistent.

```
❱ dmesg | tail
[1262076.514765] cpaCySymRemoveSession() - : There are 1 requests pending
[1262349.649144] cpaCySymRemoveSession() - : There are 1 requests pending
[1262516.514345] cpaCySymRemoveSession() - : There are 1 requests pending
[1263214.854780] cpaCySymRemoveSession() - : There are 1 requests pending
[1263687.043410] cpaCySymRemoveSession() - : There are 1 requests pending
[1264374.229766] cpaCySymRemoveSession() - : There are 1 requests pending
[1264524.037670] cpaCySymRemoveSession() - : There are 1 requests pending
[1266378.818123] cpaCySymRemoveSession() - : There are 1 requests pending
```

The system can remain in its current condition for a few days, so some troubleshooting is possible. Just let me know if there's any more information I can provide. :-)

**Comment by luki-mbi** [comment]
I ran into this couple months ago. I probably should have updated this issue, but then the world ground to a halt.

What I think happens is that the data gets corrupted during compression and thus cannot be decompressed. The checksum is fine since it's for the compressed data given to ZFS, so the scrub won't find any errors. That's bad since you won't detect this problem until you actually try to read the file and decompress the data. Decompression also failed with the software decompression without the QAT.

For me decompression failed at always the same location in the file, and every attempt to read the affected location incremented dc_fails. In the end it was just one block that had bad data; multiple large (1TB+) files were affected. I extracted the raw compressed data from the pool with zdb and tried to decompress it with gzip. No go gzip partially decompressed the input and then failed with: invalid compressed data--format violated. Basically, the compressed data is corrupt.  As a control, the procedure worked fine for adjacent blocks, i.e. immediately before and after the affected block.

I don't really know causing it but I suspect it's related to memory pressure. I probably got myself in hot water by using 1MB block size with QAT (after adjusting QAT_MAX_BUF_SIZE). So may I'm to blame for my problems. The system has 256 GB RAM and is rock solid otherwise transferring ~1PB of data per month, so I doubt it's a hardware issue.  

It would be great to track down what's really causing it. Maybe it's been fixed already. Hard to test as  I don't have a reproducer and this affected "just" 5 x 1 MB blocks over a dataset of ~400 TB.

**Comment by aqw** [comment]
I think we have the same problem, though I have not adjusted my QAT_MAX_BUF_SIZE.

Further testing showed that the file fails to decompress, with or without QAT enabled. I am able to `cat` some of the file, but then it fails (suggesting one bad block). A `zfs send` of the dataset similarly fails (as it tries to decompress the block), but succeeds with `zfs send -vcR`. So I think the block was incorrectly compressed, and there's a valid checksum of the bad data.

I have no idea what caused this. Copying the same file (from a good source) and reading it back works without error, so it's not content specific.

**Comment by AGI-admin** [comment]
In our case, we ended up getting the new QAT card replaced, but that may not have been the (only) problem.  The new card was not recognized by the system initially and may have been due to dust in the PCI slot?  Without the card recognized and ZFS compression still enabled, I was able to delete and copy data from the backup.  In our case it was important to delete the data first, otherwise I/O errors would still happen when trying to read the data again.  I guess because rsync only syncs some of the data that is different and not the entire file.  Also with the old QAT card, as described above, deleting and restoring from backup would sometimes still produce to I/O errors when trying to read the newly restored data, so maybe dust in the PCI slot was somehow causing that?

After I found time to shutdown the system again, pull the card out, blow some compressed air into the PCI slot and re-seat the card, it was detected by the system again.  The newly-restored data can be read fine now with the QAT card so we don't have anymore corrupted data or I/O errors.  So maybe due to dust in the PCI slot and/or faulty QAT card...

**Comment by aqw** [comment]
Hmmm. Thanks for the information.

We have a second system with a QAT card, and we've discovered that this problem exists there as well. It could still be hardware, but only if Intel's QA is slipping. Interestingly, the issue seems to be limited (so far) to the compression of blocks, and not decompression nor checksumming (we don't use the AES functionality yet).

We'll try the compressed air idea; it never hurts to try. ;-)

**Comment by luki-mbi** [comment]
For us, we see the same symptoms on two separate systems -- one has a PCIe card, the other one has the QAt on chipset (Supermicro X11DPH-Tq board). I doubt it's a hardware problem.

**Comment by aqw** [comment]
The thing that scares me is that these failures come from blocks that are so mangled that they fail to decompress. There is no assurance (that I'm aware of) that the other "valid" blocks actually have the correct content. It's just that they decompress without failure.

ZFS sees everything as ok, because the checksum is for the compressed block and not the content.

Are you (@luki-mbi @AGI-admin) running systems with QAT cards in production? If so, how are you testing for bad blocks, and are you doing any other form of verification that the "successful" blocks actually contain the data intended?

I can't trigger this bug on-demand, but it does come up frequently when we're running DataLad (which uses `git-annex` under-the-hood) for a few hours on large datasets. If it would be useful for a developer to see this failure up close, I can provide system access and a workload that will semi-reliably make it happen.

I don't know what layer is at fault here, but a heavy handed solution that ZFS could take is have the QAT card: checksum the content, compress, checksum the compression, decompress, compare checksum, if everything's good, write to disk. It's wasteful, but it would catch the problem, as there is no indication that the checksums generated are wrong.

**Comment by AGI-admin** [comment]
Yes we are using 2 systems in production.  One is our main server with user home directories and the other is a bacula backup server.  All of the data on the main server is decompressed and read by the bacula process and sent over the network to the bacula server where it is compressed again in a bacula volume.  Before, there were many read errors within bacula and they were all intermediary files from canu application.  Now there are no errors.  I can also check the file from above that was having problems from a second backup we still have and the checksum is matching so I don't think there are anymore problems.

**Comment by luki-mbi** [comment]
We're running both systems in production too, and while dc_fails goes up by ~30/month, it's just a tiny fraction ... like 0.1 ppm failure rate.

As of late, the system with the PCIe card works without any dc_fails but the one with the QAT on the C627 chipset does have dc_fails. I still doubt it's a hardware issue... I certainly can't re-seat the chipset :).

**Comment by ajgnet** [comment]
I am experiencing the same error with qat_c3xxx. Currently running:
zfs-0.8.4-1
zfs-kmod-0.8.4-1

`root@library:~# cat /sys/kernel/debug/qat_c3xxx_0000\:01\:00.0/cnv_errors
+-----------------------------------------------------------------+
|             CNV Error Freq Statistics for Qat Device            |
+-----------------------------------------------------------------+
|[AE  0]: TotalErrors:     6 : LastError: Decomp Error   [   -7]  |
+-----------------------------------------------------------------+
|[AE  1]: TotalErrors:    10 : LastError: Decomp Error   [   -7]  |
+-----------------------------------------------------------------+
|[AE  2]: TotalErrors:     7 : LastError: Decomp Error   [   -7]  |
+-----------------------------------------------------------------+
|[AE  3]: TotalErrors:     5 : LastError: Decomp Error   [   -7]  |
+-----------------------------------------------------------------+
|[AE  4]: TotalErrors:     9 : LastError: Decomp Error   [   -7]  |
+-----------------------------------------------------------------+
|[AE  5]: TotalErrors:    13 : LastError: Decomp Error   [   -7]  |
+-----------------------------------------------------------------+
`
And thousands of lines in my system log
`[38165.086022] cpaCySymRemoveSession() - : There are 1 requests pending
[38165.172691] cpaCySymRemoveSession() - : There are 1 requests pending
[38165.773006] cpaCySymRemoveSession() - : There are 1 requests pending
[38178.654940] cpaCySymRemoveSession() - : There are 1 requests pending
[38179.193518] cpaCySymRemoveSession() - : There are 1 requests pending
[38179.678819] cpaCySymRemoveSession() - : There are 1 requests pending
[38180.067001] cpaCySymRemoveSession() - : There are 1 requests pending
[38181.253219] cpaCySymRemoveSession() - : There are 1 requests pending
[38183.744417] cpaCySymRemoveSession() - : There are 1 requests pending
[38187.058561] cpaCySymRemoveSession() - : There are 1 requests pending
[38187.683833] cpaCySymRemoveSession() - : There are 1 requests pending
[38199.590277] cpaCySymRemoveSession() - : There are 1 requests pending
[38200.987345] cpaCySymRemoveSession() - : There are 1 requests pending
[38201.379468] cpaCySymRemoveSession() - : There are 1 requests pending
[38201.463914] cpaCySymRemoveSession() - : There are 1 requests pending
[38203.434172] cpaCySymRemoveSession() - : There are 1 requests pending
[38203.600465] cpaCySymRemoveSession() - : There are 1 requests pending
[38204.045635] cpaCySymRemoveSession() - : There are 1 requests pending
[38206.087321] cpaCySymRemoveSession() - : There are 1 requests pending
[38206.662690] cpaCySymRemoveSession() - : There are 1 requests pending
[38207.393078] cpaCySymRemoveSession() - : There are 1 requests pending
[38208.365063] cpaCySymRemoveSession() - : There are 1 requests pending
[38209.359167] cpaCySymRemoveSession() - : There are 1 requests pending
[38224.033778] cpaCySymRemoveSession() - : There are 1 requests pending
[38225.651293] cpaCySymRemoveSession() - : There are 1 requests pending
[38225.986586] cpaCySymRemoveSession() - : There are 1 requests pending
[38226.050699] cpaCySymRemoveSession() - : There are 1 requests pending
[38227.986643] cpaCySymRemoveSession() - : There are 1 requests pending
[38227.989797] cpaCySymRemoveSession() - : There are 1 requests pending`

**Comment by AGI-admin** [comment]
@ajgnet Are you having problems reading or writing data as well?  That was a critical component of the original issue so, if not, it might be another issue you are experiencing.

**Comment by ajgnet** [comment]
The data appears to write correctly; I believe the issue is with reading only. Zpool status returns no errors and zpool scrub returns no errors. This is happening on two identical C3XXXX machines w/ QAT integration. It's almost as if something goes wrong on compression yet the error isn't realized until decompression. The ZFS checksum is based on the faulty compressed data and so does not report an error until decompression

**Comment by aqw** [comment]
@ajgnet What you're experiencing sounds very similar to my case: silent corruption on compression, "correct" checksum of corrupted data (so scrubs look clean), and fail on decompression.

@luki-mbi Do your decompress fails succeed on subsequent tries? I'm wondering if you and I are experiencing different problems, or if the data you're shuffling around isn't as precious. Because you seem comfortable with the low but consistent rate of data loss, which surprises me.

**Comment by luki-mbi** [comment]
@aqw: No, as mentioned in my Dec 28, 2019 comment, the data always fails to decompress. The corruption seems to happen during compression and doesn't get caught until a decompression attempt. Neither the CPU implementation nor the QAT can decompress it. I even extracted the compressed data from the pool and attempted to manually decompress it -- see my Apr 22 comment.

So far capping the ARC size to leave free RAM is a workable workaround that triggered no errors for months, but it's not a pretty solution. We only use the QAT on archive servers with large files, and I actually fully re-read the files are writing to confirm they are OK. I really shouldn't have to.

**Comment by luki-mbi** [comment]
PS: some ZFS QAT stats since my last ZFS upgrade couple weeks ago -- no failures over ~200 TB compressed/decompressed.
```
# cat /proc/spl/kstat/zfs/qat
19 1 0x01 17 4624 1645575508916183 3447141868397572
name                            type data
comp_requests                   4    218485057
comp_total_in_bytes             4    214690002502144
comp_total_out_bytes            4    53087375405150
decomp_requests                 4    932199741
decomp_total_in_bytes           4    89658253388433
decomp_total_out_bytes          4    235924900200960
dc_fails                        4    0
encrypt_requests                4    0
encrypt_total_in_bytes          4    0
encrypt_total_out_bytes         4    0
decrypt_requests                4    0
decrypt_total_in_bytes          4    0
decrypt_total_out_bytes         4    0
crypt_fails                     4    0
cksum_requests                  4    0
cksum_total_in_bytes            4    0
cksum_fails                     4    0
```

**Comment by wli5** [comment]
@AGI-admin @luki-mbi I just saw this issue, has it been fixed in the latest ZFS release?

**Comment by wli5** [comment]
We (@cfzhu )'ve tried ZFS 0.8.4 on C62x platform, copying/reading many different files (total 30GB) works well without any issue.. will rollback to ZFS 0.8.2 to see if we can reproduce the issue.

**Comment by aqw** [comment]
@wli5 In my experience, it takes more than 30G to consistently tickle this bug. We move around 1TB and then we (usually) notice it. Just FYI.

**Comment by wli5** [comment]
@aqw are you using the latest ZFS release 0.8.4?

**Comment by aqw** [comment]
@wli5 The last one we tested with QAT was 0.8.3.

**Comment by wli5** [comment]
@aqw thanks, we will test with more data. Are you able to move to 0.8.4 to see if the issue is fixed?

**Comment by aqw** [comment]
@wli5 I have upgraded one server to 0.8.4, upgraded the QAT driver from `1.7.l.4.7.0` -> `1.7.l.4.10.0`. I am running tests again. I will probably be able to share results tomorrow.

**Comment by aqw** [comment]
An idle musing: while waiting for the tests to run, I read through resolved issues in the QAT software between 4.7 and 4.10. Issue QATE-58487 sounds related:

> QATE-58487 - DC - Compressed data fails to decompress

> If the Compress and Verify feature is explicitly disabled (which is not a supported
configuration) when performing a compression operation with Intel® QAT, with certain
input data of greater than 64KB, the compressed data cannot be decompressed. 

According to Intel's docs, CnV is enabled by default when calling `cpaDcCompressData()`, which is the endpoint used in ZFS's support for QAT compression.[1]

So QATE-58487 is probably not what was affecting us, but it made me wonder how these errors made it past QAT's CnV checks, and whether ZFS catches the recovery attempts.

[1] https://github.com/openzfs/zfs/blob/bc9cef11fddfe8e6e240477084a49f4bad1fb4f3/module/os/linux/zfs/qat_compress.c#L399

**Comment by wli5** [comment]
@aqw if the CnV check fails, ZFS code will check the result:
https://github.com/openzfs/zfs/blob/bc9cef11fddfe8e6e240477084a49f4bad1fb4f3/module/os/linux/zfs/qat_compress.c#L413

And then it will fallback to SW implementation:
https://github.com/openzfs/zfs/blob/bc9cef11fddfe8e6e240477084a49f4bad1fb4f3/module/zfs/gzip.c#L71

**Comment by aqw** [comment]
Ahh; thanks for pointing that out. My C knowledge is near-zero, so while I could see the `goto fail`, it wasn't clear to me how it was all handled. I see it now. Thanks. :-)

I have some interesting findings:

- I sent a 3 TiB dataset twice with no errors, and I can read back the resulting data cleanly.
- I have a ~220GiB dataset that /always/ generates 6 `dc_fails` (also visible in `cnv_errors`) when I send it. However, the software fallback appears to be working, as reading the data results in no errors.
- Corrupted data from before (0.8.3) is still corrupted, as expected.

The other system I have with a QAT card cannot be used for troubleshooting right now, so I cannot test the 220GB dataset on another system. The 6 `dc_fails` is reproducible every time (tested 4 times), suggesting that QAT reliably chokes on specific data. If you're interested in exploring this, I can provide you with the dataset; nothing in it is sensitive.

I'm continuing to send larger datasets, to test this further.

**Comment by luki-mbi** [comment]
This corruption still occurs sometimes, and that's really not great. For example, zfs send will reproducibly fail with "Input/output error" when it cannot decompress a block (not surprisingly) -- in the example below after 569G out of 756G. The corrupt data must have been written last month as this is an incremental send of monthly snapshots. Admittedly this server is not running the newest version of ZFS so an upgrade is probably worth a shot.

```
root@nfs-8:~# zfs send -v -c -i jbod-2/shared/xray@2020-07-01_030300 jbod-2/shared/xray@2020-08-01_030300 | ...
send from jbod-2/shared/xray@2020-07-01_030300 to jbod-2/shared/xray@2020-08-01_030300 estimated size is 756G
total estimated size is 756G
cannot receive incremental stream: checksum mismatch or incomplete stream.
Partially received snapshot is saved.
A resuming stream can be generated on the sending system by running:
    zfs send -t 1-154c7bbdfc-118-...

root@nfs-8:~# cat /proc/spl/kstat/zfs/qat
...
dc_fails                        4    1

root@nfs-8:~# zfs send -t 1-154c7bbdfc-118-... > /dev/null
warning: cannot send 'jbod-2/shared/xray@2020-08-01_030300': Input/output error

root@nfs-8:~# cat /proc/spl/kstat/zfs/qat
...
dc_fails                        4    2
```

**Comment by wli5** [comment]
@aqw thanks for the update. If possible please share the data block which can cause the dc_fails. BTW - even the block can't be compressed by HW, the SW will take over, as you pointed out: reading the data results in no errors.

**Comment by wli5** [comment]
@luki-mbi thanks for the info! Please upgrade to the latest ZFS release and see if the issue is fixed. I've tried to copy/read more than 2TB files in the whole night and no issue was found on ZFS 0.8.4.

**Comment by aqw** [comment]
@wli5 I am sad to report that this bug is not resolved. I did a large send (37 TiB); upon reading the results (also via zfs send, but redirecting to `/dev/null`), I get a decompress failure, and the send fails.

How would you like to proceed? I'm happy to assist any way I can (information, testing, etc). This is a high priority issue for us, so I'm willing to make the time if I can be of assistance.

**Comment by wli5** [comment]
@aqw have you updated to 0.8.4? can you please let me know your test method? what we did was copying many files to the pool and read them back to verify the content are same.

**Comment by aqw** [comment]
> have you updated to 0.8.4? 

@wli5 Yes, as I mentioned above, I am running ZFS 0.8.4 and QAT driver 1.7.l.4.10.0. The machine is running Debian 10 on the 4.19 kernel.

The method I did to test:
```
# a large pool with plenty of free space
$ zpool list
NAME   SIZE  ALLOC   FREE  CKPOINT  EXPANDSZ   FRAG    CAP  DEDUP   HEALTH  ALTROOT
hdd1   327T   149T   179T        -         -     1%    45%  1.00x   ONLINE  -

# enable QAT compression and checksumming for ZFS
$ echo 0 >| sys/module/zfs/parameters/zfs_qat_compress_disable
$ echo 0 >| sys/module/zfs/parameters/zfs_qat_checksum_disable

# source is gzip-9
$ zfs get compress hdd1/data1/project
NAME               PROPERTY      VALUE     SOURCE
hdd1/data/project  compressions  gzip-9    local

# create destination
$ zfs create hdd1/test/qat

# send 38T; without -C, so it will recompress on the receive
$ zfs send -vR hdd1/data1/project@2020-05-01_05.20.01--1y | zfs receive -F hdd1/test/qat

# read back the results
$ zfs send -vR hdd1/test/qat@2020-05-01_05.20.01--1y > /dev/null
[...]
warning: cannot send: 'hdd1/test/qat/digital_biomarker@2020-01-14_09.46.39--1y': Input/output error
```
- `/proc/spl/kstat/zfs/qat` shows an increase of 43 `dc_fails`
- `/sys/kernel/debug/qat_c6xx_*/cnv_errors` shows an increase of 42 failures.

I noted these numbers beforehand, so it's possible that there's a mistake here. I need to re-run the test to confirm, and capture the full contents before and after.

**Comment by aqw** [comment]
I have good news: I have a 500G dataset that I can reliably reproduce this bug with using the above method. Even after a reboot. I tested 5 times.

The point where it fails in the data stream is not consistent, suggesting that something about this data increases the probability of encountering the bug, but isn't a simple "this file/block always corrupts".

I am not sure if I can share this dataset (I need to check first), but I am happy to provide information and debugging.

**Comment by wli5** [comment]
@aqw when you saw the fail, can the files still be read? just wondering if the software fallback worked. if you can continue narrow down to a small set of data and share with us that will be helpful to debug the issue.

**Comment by aqw** [comment]
@wli5 The files with corrupted blocks can't be read... If `zfs` itself can't read the block, nothing else in the OS can.

Please re-read what I have written above. `/proc/spl/kstat/zfs/qat` and `/sys/kernel/debug/qat_c6xx_*/cnv_error` are off by 1. ZFS and QAT are both catching all of the same failures during compress (at least they end up with the same number, which is a good sign). The off-by-1 comes from reading the results (and failing). That failure is seen by ZFS but not by QAT. This makes sense, because the QAT stats are only for Compress and Verify (not for decompress alone). 

Something, somewhere, is still generating corrupted blocks. I don't know if that's a result of the fall-back failing or the worse-scenario that QAT (rarely) generates blocks that it thinks are valid, but actually aren't.

I also tried @luki-mbi's suggestion to limit the ARC, as it helped them. I capped at 50% (leaving ~128GB free) and the problem is sadly still reproducible.

I will see if I can get a smaller dataset that triggers this problem. I was already excited that I could reliable trigger this with just 500GB, as it's such a low-frequency problem.

My suspicion is that this problem happens with many small files. The dataset that reliably triggers this bug is 1.5 million JSON files totaling ~500GB (average size ~330KB).

**Comment by wli5** [comment]
@aqw are you able to reduce the file size?

**Comment by aqw** [comment]
@wli5 I have not yet been able to generate a smaller dataset that tickles this. I'm still looking.

**Comment by aqw** [comment]
I came back to this, and I can only consistently reproduce it with that 500GB dataset. Other workloads do still cause this problem, but I can't get the reliability of failure up.

@wli5 What information can I provide that will help you forward on this? Currently, all of our QAT cards remain idle due to this bug.

**Comment by wli5** [comment]
@aqw is it possible to figure out which piece of data caused the issue? Is it on the same block or random? Can you just test part of the data within the 500GB to see which part caused the issue?

**Comment by aqw** [comment]
It's not always the same block (oddly enough), but is definitely not random. It's usually around similar blocks.

I don't know how to use `zdb` well enough to inspect the failed blocks themselves. Any tips?

**Comment by wli5** [comment]
@behlendorf maybe Brian can give some advise on the zdb?

**Comment by wli5** [comment]
@aqw is the 500GB a single file or multiple files? Can you narrow down which file caused the issue?

**Comment by aqw** [comment]
@wli5 I have provided much of this info before (see above).

The 500G is the size of the dataset. I am primarily testing this by using `zfs send` to send the entire dataset. I do this, because the dataset is 1.5 million files (average size ~330KB). Using `rsync` also encounters the bug, but I use `zfs send` because it's faster than `rsync`.

I have isolated the files it fails on (simple JSON files). Sending just those does not cause the bug.

This bug behaves as if the QAT card needs to be primed with a certain type of data, then another type of data causes corrupted output. That's my best guess.

**Comment by wli5** [comment]
@aqw I see.. thanks for the info. Can your files be shared so we can reproduce in our lab?

**Comment by wli5** [comment]
Hi @meckiemac thanks! Can you please share more details how to reproduce the issue?

**Comment by meckiemac** [comment]
In my case it's my home server Supermicro SYS-E200-9A. The system is on an NVMe in the original NVMe slot and the user data resides on another M2 NVMe via an adapter in the PCIe slot. A third HDD is connected via SATA as backup disk.
Both data NVMe and backup SATA drive are on ZFS, the system NVMe runs ext4 via lvm.  Further it has 32GB ECC RAM and zfs scrub is clean and zfs arc size was default 1/2 RAM, but now reduced to 8GB (after the errors).
The error was found during backups via syncoid from the data to the backup medium.

I don't know when the error was introduced, but it seem reproducible. The virtual machine experiencing the issue got created with virt-install and the disk size (qcow2) was internal 30GB. Between creation and backup there was 2-4 hours the last time I experienced it. The high count of the error comes from the retry of the disk subsystem in the virtual machine.

With its 4 cores the atom is sometimes under good load. I have also an nextcloud instance running and so on.

So the reproducible step is: create a virtual machine with windows 10 as guest … and wait few hours of write activities in my case. (It happened before, but with a longer time between creation and detection.)

**Comment by wli5** [comment]
Thanks @meckiemac , but I don't quite understand. You have Linux machine, and have ZFS running on it for an NVMe disk and a SATA drive, and you have Windows10 virtual machine on this Linux server... then how to reproduce the issue, as you said during backups, how? From windows VM?

**Comment by meckiemac** [comment]
* To reproduce the issue I just create ta Windows10 virtual machine (had this with XP too) and/or let it run for a short time. The virtual machines running on the nvme. This was enough to produce corruption.
* The detection of the corruption was during a syncoid backup of the filesystem to the backup disk (all zfs). After detection any read of the file containing the vm failed, inside or outside read (dd, windows disk backup).

Currently I disabled gzip compression on large files and virtual machines. I also do regular zfs send to null to check the rest of the data. So far the issue only found with virtual machine disk files (qcow2)

Does this help better to understand the scenario?

**Comment by wli5** [comment]
@merkiemc Hi Andreas, thanks for the info! If you write/read the qcow2 file to ZFS with compression, can the issue be triggered?

**Comment by meckiemac** [comment]
I haven't tried to trigger the corruption on purpose yet. The system is also my home "production" system, so I'm a bit careful. I found another compression corruption on the Sata disk (Backup) in the OSX-"Time Machine Backup" space. The "Time Machine Backup" space is provided via Samba to My OSX clients. The big difference here is that we have a lot of smaller files compared to the virtual machines. Currently I'm trying to clear this corruption, since the Time Machine is very sensitive with disk errors and finally also switch off compression here. 

What I can try to setup is a script which is installing windows virtual machines and then checks the written data in a loop, until we trigger the corruption, since this is the most obvious area where this error occurred. With the script, if successful, we could try to collect more data or modify the system and observe the behavior.

**Comment by wli5** [comment]
If you can provide the script it can also help us to reproduce the issue. Thanks!

**Comment by wli5** [comment]
We have tried to reproduce this issue by creating VM and backup snapshot regularly to a compressed ZFS pool (QAT enabled), by running more than 2 weeks repeatly, without any issue.

**Comment by meckiemac** [comment]
@wli5 my process would looks exactly similar, but I haven't found the time to implement it yet. Family and job have currently priority. The interesting part: it is not only bound to VMs since I have the same corruption in backup data thru a Samba-server. So I guess it is connected to the system load too. To be able to automate my VM I need to create an unattended setup for windows 10 first, which takes some time. Sorry for the delay. The current workaround for me is switching of compression. The archieved compression rate is negligible compared to my available space.
